### DISCUSSION TITLE
Is it possible to implement matching XML start/end tag names without an external scanner?

### DISCUSSION THREAD

**Discussion by s1341** [tree-sitter/tree-sitter#3152] [Q&A] [answered]
Hi,

I'm interested in writing a rule to generate an XML element, with matching start and end tag names.

Something like:
```js
    element: $ => {
      var n = $.Name;
      return choice(
        $.EmptyElemTag,
        seq(
          seq(
            '<',
            n,
            c.rseq($._S, $.Attribute),
            O($._S),
            '>'
          ), O($.content),
          seq(
            '</',
            n,
            O($._S),
            '>'
          )
        )
      )
    },
```

But I'd like `n` to be instantiated once for the rule. I.e. to run the `$.Name` rule once. Is there a way to do this in tree-sitter?

**Comment by amaanq** [comment]
To answer your question directly, no, there is no way to do so without an external scanner, that requires being context aware. However, have you seen https://github.com/tree-sitter-grammars/tree-sitter-xml? It's implemented there, but did require the usage of an external scanner.

**Reply by s1341** [reply]
I did see it, which is why I was asking if there is a way to do it without an external scanner ;)

**Reply by amaanq** [reply, marked as answer]
yeah nah sadly, it'd be really hard to represent in the dsl and in the parser
